src/main/scala/com/databricks/labs/overwatch/pipeline/SilverTransforms.scala

```diff
@@ -1314,4 +1314,119 @@ trait SilverTransforms extends SparkSessionWrapper {
+      'spot_instance_policy,
+      'enable_photon.cast("boolean"),
+      'channel,
+      'tags,
+      'enable_serverless_compute.cast("boolean"),
+      'warehouse_type
+    )
+
+    val warehouseRaw = auditRawDF
+      .filter('serviceName === "databrickssql")
+      .selectExpr("*", "requestParams.*").drop("requestParams", "Overwatch_RunID")
+      .select(warehouseSummaryCols: _*)
+      .withColumn("warehouse_id", warehouse_id_gen)
+      .withColumn("warehouse_name", warehouse_name_gen)
+
+    val warehouseWithStructs = warehouseRaw
+      .withColumn("channel", SchemaTools.structFromJson(spark, warehouseRaw, "channel"))
+      .withColumn("tags", SchemaTools.structFromJson(spark, warehouseRaw, "tags"))
+      .scrubSchema
+
+    warehouseWithStructs
+        .withColumn("tags", SchemaTools.structToMap(warehouseWithStructs, "tags"))
+  }
+
+  protected def buildWarehouseSpec(
+                                  bronze_warehouse_snap: PipelineTable,
+                                  isFirstRun: Boolean,
+                                  untilTime: TimeTypes
+                                )(df: DataFrame): DataFrame = {
+
+    val warehouseBaseDF = warehouseBase(df)
+    val warehouseBaseWMetaDF = warehouseBaseDF
+      // remove start, startResults, and permanentDelete as they do not contain sufficient metadata
+      .filter('actionName.isin("createEndpoint", "editEndpoint", "createWarehouse", "editWarehouse"))
+    val bronzeWarehouseSnapUntilCurrent = bronze_warehouse_snap.asDF
+      .filter('Pipeline_SnapTS <= untilTime.asColumnTS)
+
+    /**
+     * warehouseBaseFilled - if first run, baseline warehouse spec for existing warehouses that haven't been edited since
+     * commencement of audit logs. Allows for joins directly to gold warehouse work even if they haven't yet been edited.
+     * Several of the fields are unavailable through this method but many are and they are very valuable when
+     * present in gold
+     */
+    val warehouseBaseFilled = if (isFirstRun) {
+      val firstRunMsg = "Silver_WarehouseSpec -- First run detected, will impute warehouse state from bronze to derive " +
+        "current initial state for all existing warehouses."
+      logger.log(Level.INFO, firstRunMsg)
+      println(firstRunMsg)
+      val missingWarehouseIds = bronzeWarehouseSnapUntilCurrent.select('organization_id,
+        'warehouse_id).distinct
+        .join(
+          warehouseBaseWMetaDF
+            .select('organization_id, 'warehouse_id).distinct,
+          Seq("organization_id", "warehouse_id"), "anti"
+        )
+      val latestWarehouseSnapW = Window.partitionBy('organization_id, 'warehouse_id).orderBy('Pipeline_SnapTS.desc)
+
+      val missingWareHouseBaseFromSnap = bronzeWarehouseSnapUntilCurrent
+        .join(missingWarehouseIds, Seq("organization_id", "warehouse_id"))
+        .withColumn("rnk", rank().over(latestWarehouseSnapW))
+        .filter('rnk === 1).drop("rnk")
+        .select(
```

Comment:
you're force-casting several of these types from the audit logs -- be sure the types match on all the fields. I'm guessing you've already verified this via the WarehouseSnap Schema, just want to make sure.